Add option to configure datastore_cluster

dirkmueller opened the pull request:
Using 3PAR over network ISCSI is not a great idea regarding IOPS
performance which matters for etcd.  Add ability to set a datastore_cluster
as well (optionally)

## Why is this PR needed?

It fixes the e2e CI failures on vmware because they are currently using
ISCSI (which is very slow) on 3PAR (which is extremely slow). 

Fixes #

**Reminder**: Add the "fixes bsc#XXXX" to the title of the commit so that it will
appear in the changelog.

## What does this PR do?

please include a brief "management" technical overview (details are in the code)

## Anything else a reviewer needs to know?

Special test cases, manual steps, links to resources or anything else that could be helpful to the reviewer.

## Info for QA

This is info for QA so that they can validate this. This is **mandatory** if this PR fixes a bug.
If this is a new feature, a good description in "What does this PR do" may be enough.

## Docs

Documentation update is in https://github.com/SUSE/doc-caasp/pull/629


# Merge restrictions

(Please do not edit this)

We are in *v4-maintenance phase*, so we will restrict what can be merged to prevent unexpected surprises:

    What can be merged (merge criteria):
        2 approvals:
            1 developer: code is fine
            1 QA: QA is fine
        there is a PR for updating documentation (or a statement that this is not needed)

<!-- Remember, if this is a work in progress please pre-append [WIP] to the title until you are ready! 
    If you can, please apply all applicable labels to help reviews out! -->

dirkmueller commented:
ok, I see this is not so simple potentially - https://github.com/lcavajani/skuba/commit/2f4e58db951214cb7d4add1220baab4227795191

pablochacin commented:
The problem I see is that even when those changes are simple we still need to find a way to automate the way we introduce it without changing the terraform files.  We could simply prepare a patch and apply it before running the test.

thehejik commented:
> The problem I see is that even when those changes are simple we still need to find a way to automate the way we introduce it without changing the terraform files. We could simply prepare a patch and apply it before running the test.

Exactly, we are doing it on the fly within validation by using sed and echo but patch would work as well.

dirkmueller commented:
thansk for setting this to "needs change" without actually explaining what needs change. With VMware you have (at least) 3 options on how to allocate storage for a virtual machine:

* using VMFS, which afaik is the default
* using a clustered datastore (which is a internal datastore clustering inside vmware, similar to e.g. ceph in SUSE world)
* using a hardcoded datastore (which could be a particular local drive or a NAS drive). 

as far as I can see, limiting to a datastore seems to be the least versatile option to use for deployment, and in this case is outright harmful because it is an iSCS attached drive which is just too slow for the rest of the system's expectations. 

The right choice would be a) to make the ci default to a local disk and b) give the user options of deployment. 

Shipping terraform files that we're not using ourselves and instead letting QA sed the terraform files before validating the product seems against the principle of testing what we ship.

pablochacin commented:
> thansk for setting this to "needs change" without actually explaining what needs change.  

I think @thehejik was pretty clear, pointing to the actual changes that need to be made. 

> Shipping terraform files that we're not using ourselves and instead letting QA sed the terraform files before validating the product seems against the principle of testing what we ship.

Agree, but introducing this change as a temporary measure in order to have the tests running until we fix the problem with slow disks seems better than not testing anything at all, IMHO.

dirkmueller commented:
@pablochacin I was mentioning the patch already in previous comment. 

So I just submit the patch that was referenced and that we already decided against merging? Why?

Ar least it would make sense to make the storage backend selectable instead of preferring the least useful option. Or like this patch does, not set it at all.. which makes it chose some default.

jordimassaguerpla commented:
> @pablochacin I was mentioning the patch already in previous comment.
> 
> So I just submit the patch that was referenced and that we already decided against merging? Why?
> 
> Ar least it would make sense to make the storage backend selectable instead of preferring the least useful option. Or like this patch does, not set it at all.. which makes it chose some default.

IIUC our customers use our tf files as a template. Not choosing a default can make sense as long as we update the documentation to reflect this change.

If we don't use a default, will we be using the local storage in vmware? or do we need to change that?

dirkmueller commented:
@jordimassaguerpla yes, afaik it will use vmfs, which is defaulting to whatever has been configured by the admin (local datastores for example)

Anyway, I've updated the PR to include the ability to set the datastore_cluster in addition to the datastore, and changed the CI worker to set the datastore_cluster rather than datastore. I hope this is a sensible compromise between setting datastores (still not sure why we would want to do that but it has been the default so far) and datatore_cluster (which we need for CI).

jordimassaguerpla commented:
@dirkmueller I retriggered the tests and they are green now. Please see the requested changes to get this approved.Thanks.

dirkmueller commented:
@jordimassaguerpla  we need to do this after the TF 0.12 switch

jordimassaguerpla commented:
Hi!
New code has been merged in master regarding terraform 0.12. Please rebase your branch to take this changes in.
Happy hacking

evrardjp commented:
Just curious, how does this relate to https://github.com/SUSE/skuba/pull/950/files ?

dirkmueller commented:
@evrardjp not at all, this pr is a month older. 

The original attempt was at making the terraform files usable so that you can configure either datastore or datastore_cluster (the latter I think being more relevant to customers, but YMMV). 

The other PR switches to a different cluster on better hardware which might be a quick fix to the current stability issues. It is incremental. 

Note that this PR is currently not in a mergeable state as terraform fails with that change and I'm slow with debugging this because I have to debug it via CI as I don't have credentials to test this outside CI. 

Please do not merge it until I got the job to pass or at least a way to fix the regression.

evrardjp commented:
Hey, if I merge I will break world, I am not looking forward to that :p 
My question was more about -- "Are you aware of this relatively new PR? Do we still need this?"

You answer tells me that datastore_cluster is more relevant for customers, which explains why it's not closed.

Itxaka commented on ci/infra/vmware/terraform.tfvars.json.ci.example:
```diff
@@ -1,6 +1,7 @@
 {
     "vsphere_datastore": "3PAR",
     "vsphere_datacenter": "PROVO",
+    "vsphere_datastore_cluster": null,
```
I think deployment is failing because this is set to null. Should be set to `LOCAL-DISKS-CLUSTER` for our current state

Itxaka replied:
now that defautls are set to null we can override this for our CI and remove `vsphere_datastore` var from here :D

Itxaka commented on ci/infra/vmware/worker-instance.tf:
```diff
@@ -67,6 +67,7 @@ resource "vsphere_virtual_machine" "worker" {
   scsi_type        = data.vsphere_virtual_machine.template.scsi_type
   resource_pool_id = data.vsphere_resource_pool.pool.id
   datastore_id     = data.vsphere_datastore.datastore.id
+  datastore_cluster_id = data.vsphere_datastore_cluster.datastore.id
```
cant use both. If `datastore_id` and `datastore_cluster_id` is set at the same time, deployment fails:
`Error: "datastore_cluster_id": conflicts with datastore_id`

Itxaka commented:
FYI this is needed to use the datastore cluster in NUE as there is no shared storage or anything in there to use generic names https://github.com/SUSE/avant-garde/issues/1354

jordimassaguerpla commented:
Will this have any impact on our customers? Anything we should mention in the Release Notes?

Itxaka commented:
> Will this have any impact on our customers? Anything we should mention in the Release Notes?

Should not have an impact but we should mention that we can now choose to use a datastore_cluster **and** document that the options are exclusive, either you choose a datastore or a datastore_cluster.

jordimassaguerpla commented:
> > Will this have any impact on our customers? Anything we should mention in the Release Notes?
> 
> Should not have an impact but we should mention that we can now choose to use a datastore_cluster **and** document that the options are exclusive, either you choose a datastore or a datastore_cluster.

Ok. Then we will need also a PR to documentation and/or Release notes in order to merge this. It does not need to be a perfect PR, can be a WIP, but just so we not forget.

I'll trust you will do that **before** merging, so I am giving my LGTM

Itxaka commented:
> Ok. Then we will need also a PR to documentation and/or Release notes in order to merge this. It does not need to be a perfect PR, can be a WIP, but just so we not forget.

Rocke is borked so asking here. Do we have lables to add to PR for docs-required/release-notes-required? Should it be part of the card acceptance criteria? Both?

jordimassaguerpla commented:
> > Ok. Then we will need also a PR to documentation and/or Release notes in order to merge this. It does not need to be a perfect PR, can be a WIP, but just so we not forget.
> 
> Rocke is borked so asking here. Do we have lables to add to PR for docs-required/release-notes-required? Should it be part of the card acceptance criteria? Both?

you need an issue in github/SUSE/avant-garde. In the description of that issue, there is a section at the bottom for related issues/urls ... you can put there both, this PR and the PRs for the docs

dirkmueller commented:
The documentation has been updated already many months ago.

See the issue linked at the beginning of the PR

Itxaka commented:
>The documentation has been updated already many months ago.
>See the issue linked at the beginning of the PR

but that was reverted! https://github.com/SUSE/doc-caasp/pull/663

Plus my docs are cooler, they have a big IMPORTANT note in there about exclusivity of the options :p
